filetracker/servers/storage.py

```diff
@@ -0,0 +1,257 @@
+  have their own modification time ("version") different from the
+  modification time of the blob.
+
+- Additional metadata about blobs is stored in a BSDDB kv-store.
+- The only metadata stored ATM is the symlink count.
+- Accesses to DB are protected by fcntl locks, with one lock
+  for each prefix (256 total).
+"""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import errno
+import fcntl
+import gzip
+import hashlib
+import os
+import shutil
+import tempfile
+
+import bsddb3
+
+
+class FileStorage(object):
+    """Manages the whole file storage."""
+
+    def __init__(self, base_dir):
+        self.base_dir = base_dir
+        self.blobs_dir = os.path.join(base_dir, 'blobs')
+        self.links_dir = os.path.join(base_dir, 'links')
+        self.locks_dir = os.path.join(base_dir, 'locks')
+
+        self.db_path = os.path.join(base_dir, 'metadata.db')
+        self.db = bsddb3.hashopen(self.db_path)
+
+        # Maps prefixes to lock FDs.
+        self._db_locks = {}
+
+        _makedirs(self.blobs_dir)
+        _makedirs(self.links_dir)
+        _makedirs(self.locks_dir)
+        _makedirs(self.db_path)
+
+    def store(self, name, data, version, size=0, compressed=False, digest=None):
+        """Adds a new file to the storage.
+        
+        If the file with the same name existed before, it's not
+        guaranteed that the link for the old version will exist until
+        the operation completes, but it's guaranteed that the link
+        will never point to an invalid blob.
+
+        Args:
+            name: name of the file being stored.
+                May contain slashes that are treated as path separators.
+            data: binary file-like object with file contents. 
+                Files with unknown length are supported for compatibility with
+                WSGI interface: ``size`` parameter should be passed in these
+                cases.
+            version: new file "version"
+                Link modification time will be set to this timestamp. If
+                the link exists, and its modification time is higher, the
+                file is not overwritten.
+            size: length of ``data`` in bytes
+                If not 0, this takes priority over internal ``data`` size.
+            compressed: whether ``data`` is gzip-compressed
+                If True, the compression is skipped, and file is written as-is.
+                Note that the current server implementation sends
+                'Content-Encoding' header anyway, mandating client to
+                decompress the file.
+            digest: SHA256 digest of the file before compression
+                If specified, the digest will not be computed again, saving
+                resources.
+        """
+        link_path = self._link_path(name)
+        if os.path.exists(link_path) and _file_version(link_path) > version:
```

Comment:
Skoro nie zablokowaliśmy tutaj jeszcze tej ścieżki, to może się zdarzyć, że pomiędzy wywołaniem `exists` a `_file_version`, ścieżka zostanie skasowana.

Reply:
Racja, preniosłem wewnątrz locku.